CIPs/cip-0032.md

```diff
@@ -13,81 +13,105 @@ license: Apache 2.0
+Despite Validator uptime on Celo being phenomenal, the maintenance of the AS nodes are [sometimes neglected](https://metabase.celo-networks-dev.org/public/dashboard/b0a27650-1d62-4645-81d7-26ff7546ff0d?date_filter=past2weeks~). Today the implicit motivation to run an AS is to contribute to the health of the network which can help procure more votes from those that are interested in the success of Celo. This proposal leverages the completion rate of an Attestation Service to determine the payout percentage that a validator will receive.
 
-Rather than introducing a slashing policy which punishes poor performance, there could be an additional reward to validators for great AS performance. This option hasn't been ruled out and will likely be proposed once there are more granular metrics (see [Using completion rate](#using-completion-rate)). Today, many validators put a good amount of effort into maintaining a highly available Attestation Service. Slashing based on poor performance ensures that everyone is putting in this same level of effort to grow our network for the same compensation.
+## Specification
 
-In order to slash effectively, the contract will need an on-chain measure of AS availability. We're researching ways to provide this measurement in a manner that ensures validators are correctly rewarded when their AS is healthy and correctly punished when their AS is down. Until Celo has a large number of users, it's difficult to isolate a signal we can be confident in as a community. Therefore, this proposal is starting with a conservative approach of measuring a minimum completion threshold over a period of time.
+The `Validators.distributeEpochPaymentsFromSigner()` method will be updated to take into account the status and performance of the signer's Attestation Service. The `maxPayment` will be divided into two payments, with 80% going to the operation of the Validator and 20% going to the operation of the Attestation Service. The 80% will act as the new base reward for the validator (`actual reward = base * score * slashingMultiplier`).
 
-## Specification
+The Attestation Reward will be affected directly by the completion rate of the Attestation Service during each epoch. Since a 100% completion rate is unrealistic, a governable target completion rate will be used to determine the rewards. To ensure that an Attestation Service isn't incorrectly penalized by a bug outside of their responsibilities (ex. Valora), a minimum global attestation completion rate for this epoch will be first considered. If the minimum global attestation completion rate is not met, all elected Attestation Services will be rewarded with 75% the maximum amount, regardless of their score. A registered Attestation Service can be forcibly deregistered if they don't meet the minimum number of attestations over a governable sliding window. The Attestation Reward will be forfeited if the validator does not have a registered Attestation Service.
```

Comment:
>  a minimum global attestation completion rate for this epoch will be first considered

I'd love for us to consider automatic mechanisms for setting this rate. I am worried that any static rate would run the risk of unfairly penalizing all validators in case of coordinated failure, such as a natural disaster (e.g. hurricane, military coup, Twilio outage, etc...) or bug in an attestation client (e.g. Valora) that causes attestations to fail.

One example would be to take the mean completion rate and define the threshold as some fraction of it. (e.g. set the fraction to `1/3`) This would prevent an attacker or coordinated failure from bring down the rewards of all validators. This fraction would, by [Markov's inequality](https://en.wikipedia.org/wiki/Markov%27s_inequality), define an upper bound on the number of validator's that can fall below it at any given time. (It would not however, prevent an attacker motivated to _prevent_ a particular subset of validators from losing rewards)

Reply:
Love the idea @nategraf! In this case, we could set the threshold as 1/3 of the moving average over the last n epochs? I  like how it would auto increase/decrease the threshold as things outside of the AS node's control changed. In the case of an abrupt, sustained outage (over n epochs), the threshold will gradually drop leading to reduced payouts. I suppose if this behavior were not desirable, we could always lower the target completion rate via governance; however the time-delayed penalty seems like a benefit as it allows everyone enough time to recover before penalizing. WDYT?

Reply:
@nategraf updated the docs to reflect this. I chose a 30 day sliding window to measure the baseline to ensure that a multi-day outage doesn't pull the baseline down too much.